app/src/main/java/com/nowak/wjw/simplecompass/ui/main/MainFragment.java

```diff
@@ -71,54 +54,26 @@ public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup c
         mBinding.setLifecycleOwner(getViewLifecycleOwner());
 
         updateValuesFromBundle(savedInstanceState);
-        Timber.d("isRequestingLocationUpdates  :%s", isRequestingLocationUpdates);
-
-        AppContainer appContainer = ((MyApplication) requireActivity().getApplication()).appContainer;
-        mViewModel = new ViewModelProvider(this).get(MainViewModel.class);
+        MainViewModelFactory mainViewModelFactory = ((MyApplication) getActivity().getApplication()).appContainer.mainViewModelFactory();
+        mViewModel = new ViewModelProvider(this, mainViewModelFactory).get(MainViewModel.class);
         mBinding.setViewModel(mViewModel);
-        mViewModel.getmFoundLastLocation().observe(getViewLifecycleOwner(), is -> {
-            if (!is.getHasBeenHandled()) {
-                Timber.d("vm.foundLocation: %s", is.peekContent());
-                foundLastLocation = is.getContentIfNotHandled();
+        mViewModel.needScreenOrientation.observe(getViewLifecycleOwner(), isNeeded -> {
+            if (isNeeded) {
+                int mScreenRotation = requireActivity().getWindowManager().getDefaultDisplay().getRotation();
+                mViewModel.provideScreenRotation(mScreenRotation);
             }
         });
-        Timber.d("foundLastLocation  :%s", foundLastLocation);
+
         mViewModel.hideKeyBoard.observe(getViewLifecycleOwner(), hide -> {
             if (hide) {
                 hideKeyboard();
             }
         });
 
-        mViewModel.stopLocationUpdates().observe(getViewLifecycleOwner(), s -> {
-            Timber.d("stoplocationUpdates: %s", s);
-            if (s) stopLocationUpdates();
-        });
-
-        mViewModel.startLocationUpdates().observe(getViewLifecycleOwner(), s -> {
-            Timber.d("startlocationUpdates: %s", s);
-            if (s) requestLocationUpdates();
-        });
-
-
-        //azimuth feature
-        //todo find out scope of sensorManager
-        mSensorManager = appContainer.mSensorManager;
-        mVectorRotationSensor = appContainer.mVectorRotationSensor;
-
-        //location feature
-        mLocationApiHandler = new LocationApiHandler(LocationServices.getFusedLocationProviderClient(getContext()));
-        mLocationApiHandler.getLocation().observe(getViewLifecycleOwner(), l -> {
-//            Timber.d("newlocation");
-            mViewModel.locationChanged(l);
-        });
-
-
         mBinding.findBt.setOnClickListener((v) -> {
             if (checkPermissions()) {
                 Timber.d("button clicked & permission had been granted");
-                isRequestingLocationUpdates = true;
-                if (!foundLastLocation) configureLocation();
-                findButtonClicked();
+                btnClickAllowed();
```

Comment:
I would like to deal with the buttonClicked directly in ViewModel, but I cannot figure out how to do that.
VM registers the click and need to check if permission granted.
If yes that's awesome, easy.
But if not, he needs to notify fragment about that ,and later if permission granted, the vm.buttonClicked has to be called once again with appropriate params - I don't know how to handle it